apps/webapp/src/components/swap/asset-out-box.tsx

```diff
@@ -0,0 +1,55 @@
+import { useStore } from '../../state';
+import { swapSelector } from '../../state/swap.ts';
+import { AccountBalance, AssetBalance, groupByAsset } from '../../fetchers/balances.ts';
+import { Input } from '@penumbra-zone/ui';
+import { AssetOutSelector } from './asset-out-selector.tsx';
+import { fromBaseUnitAmount } from '@penumbra-zone/types';
+
+interface AssetOutBoxProps {
+  balances: AccountBalance[];
+}
+
+export const AssetOutBox = ({ balances }: AssetOutBoxProps) => {
+  const { assetOut, setAssetOut } = useStore(swapSelector);
+
+  const aggregatedBalances = balances
+    .flatMap(b => b.balances)
+    .reduce<AssetBalance[]>(groupByAsset, []);
+
+  const balanceOfDenom = aggregatedBalances.find(b =>
+    b.assetId.equals(assetOut?.metadata.penumbraAssetId),
+  );
```

Comment:
I'm personally a fan of doing as little logic in the template as possible, so I'd prefer adding another const below this one, something like this:
```suggestion
  );
  const formattedBalanceOfDenom = balanceOfDenom
              ? fromBaseUnitAmount(balanceOfDenom.amount, balanceOfDenom.denom.exponent).toFormat()
              : '0';
```

Then, you can just render `{formattedBalanceOfDenom}` in the markup below.

(Not a blocker though — just a nitpick.)

Reply:
why is this component manually formatting the balances, instead of having that logic only in one place, in a component rendering a `ValueView` (the same way we only format addresses once in a component rendering an `AddressView`)?